proving/proving.go

```diff
@@ -3,326 +3,165 @@ package proving
+	options := &option{}
+	for _, opt := range opts {
+		if err := opt(options); err != nil {
+			return nil, nil, err
+		}
 	}
-
-	if err := p.verifyGenerateProofAllowed(m); err != nil {
+	if err := options.validate(); err != nil {
 		return nil, nil, err
 	}
 
-	numLabels := uint64(m.NumUnits) * p.cfg.LabelsPerUnit
+	batchChan := make(chan *batch)
+	solutionChan := make(chan *solution)
 
-	for i := 0; i < MaxNumIterations; i++ {
-		startNonce := uint32(i) * NumNoncesPerIteration
-		endNonce := startNonce + NumNoncesPerIteration - 1
+	numLabels := uint64(options.numUnits) * uint64(cfg.LabelsPerUnit)
 
-		p.logger.Debug("proving: starting iteration %d; startNonce: %v, endNonce: %v, challenge: %x", i+1, startNonce, endNonce, challenge)
-
-		solutionNonceResult, err := p.tryNonces(numLabels, challenge, startNonce, endNonce)
-		if err != nil {
-			return nil, nil, err
-		}
+	workerCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	eg, egCtx := errgroup.WithContext(workerCtx)
+	eg.Go(func() error {
+		return ioWorker(egCtx, batchChan, cfg.B, options.dataSource)
+	})
 
-		if solutionNonceResult != nil {
-			p.logger.Info("proving: generated proof after %d iteration(s)", i+1)
+	var wg sync.WaitGroup
+	for i := 0; i < NumWorkers; i++ {
+		wg.Add(1)
+		eg.Go(func() error {
+			defer wg.Done()
+
+			difficulty := shared.ProvingDifficulty2(numLabels, cfg.B, cfg.K1)
+			logger.Debug("proving difficulty: %d", difficulty)
+			d := shared.CalcD(numLabels, cfg.B)
+			numOuts := uint8(math.Ceil(float64(cfg.N) * float64(d) / aes.BlockSize))
+			return labelWorker(egCtx, batchChan, solutionChan, ch, numOuts, cfg.N, cfg.B, d, difficulty)
+		})
+	}
+
+	result := &nonceResult{}
+	eg.Go(func() error {
+		var err error
+		result, err = solutionWorker(egCtx, solutionChan, numLabels, cfg.K2, logger)
+		cancel()
+		return err
+	})
 
-			proof := &Proof{
-				Nonce:   solutionNonceResult.nonce,
-				Indices: solutionNonceResult.indices,
-			}
-			proofMetadata := &ProofMetadata{
-				NodeId:          p.nodeId,
-				CommitmentAtxId: p.commitmentAtxId,
-				Challenge:       challenge,
-				BitsPerLabel:    p.cfg.BitsPerLabel,
-				LabelsPerUnit:   p.cfg.LabelsPerUnit,
-				NumUnits:        m.NumUnits,
-				K1:              p.cfg.K1,
-				K2:              p.cfg.K2,
-			}
-			return proof, proofMetadata, nil
-		}
+	wg.Wait()
+	close(solutionChan)
+	if err := eg.Wait(); err != nil && err != context.Canceled {
+		return nil, nil, err
 	}
 
-	return nil, nil, fmt.Errorf("failed to generate proof; tried %v iterations, %v nonces each", MaxNumIterations, NumNoncesPerIteration)
-}
-
-func (p *Prover) SetLogger(logger Logger) {
-	p.logger = logger
-}
-
-func (p *Prover) verifyGenerateProofAllowed(m *Metadata) error {
-	if err := p.verifyMetadata(m); err != nil {
-		return err
+	select {
+	case <-ctx.Done():
+		return nil, nil, ctx.Err()
+	default:
```

Comment:
Not necessarily. The `Generate` function receives a context from the caller, which the caller can cancel in case they are not interested in the result any more. This ends all workers, causing `eg.Wait()` to return on line 81, and we need to check here and also return.

Additionally the workers have an `errGroup` context derived from the `Generate` context which they use to signal each other to stop working when they found a solution. If `solutionWorker` found enough indices for a proof it calls `cancel` on the `errGroup` context, causing `ioWorker` and `labelsWorker` to stop and return. This is why `err == context.Cancelled` is not considered an error on line 81, since it may have been caused by the caller of `Generate`, or by `solutionWorker`. This check here is necessary to distinguish between those two cases.

Essentially if `ctx.Done()` doesn't block we don't have a proof, but if it does we have one.